src/server/dflycmd.cc

```diff
@@ -341,29 +432,25 @@ void DflyCmd::StopFullSyncInThread(FlowInfo* flow, EngineShard* shard) {
   }
 
   // Wait for full sync to finish.
-  if (flow->full_sync_fb.joinable()) {
-    flow->full_sync_fb.join();
+  if (flow->full_sync_fb.IsJoinable()) {
+    flow->full_sync_fb.Join();
   }
 
   // Reset cleanup and saver
   flow->cleanup = []() {};
   flow->saver.reset();
 }
 
-OpStatus DflyCmd::StartStableSyncInThread(FlowInfo* flow, EngineShard* shard) {
-  // Register journal listener and cleanup.
-  uint32_t cb_id = 0;
+OpStatus DflyCmd::StartStableSyncInThread(FlowInfo* flow, Context* cntx, EngineShard* shard) {
+  shared_ptr<JournalStreamer> streamer;
   if (shard != nullptr) {
-    JournalWriter writer{flow->conn->socket()};
-    auto journal_cb = [flow, writer = std::move(writer)](const journal::Entry& je) mutable {
-      writer.Write(je);
-    };
-    cb_id = sf_->journal()->RegisterOnChange(std::move(journal_cb));
+    streamer.reset(new JournalStreamer(flow->conn->socket(), sf_->journal(), cntx));
+    streamer->Start();
   }
 
-  flow->cleanup = [flow, this, cb_id]() {
-    if (cb_id)
-      sf_->journal()->UnregisterOnChange(cb_id);
+  flow->cleanup = [this, flow, streamer = std::move(streamer)]() {
+    if (streamer)
+      streamer->Cancel();
     flow->TryShutdownSocket();
```

Comment:
Storing the streamer in the cleanup callback and declaring it only inside the source file hides it from the flow and makes it harder to follow, but currently there is no reason to show it in the header.